Support local android builds for libgav1

vigneshvg opened the pull request:
When building libavif for android using the NDK, the underlying
AV1 codec can be built for multiple architectures with specific
optimizations. It is useful to build them as it is sometimes
necessary to generate a fat APK (an APK that contains the library
for multiple architectures).

Upate the CMake file to include an architecture sub-directory while
finding libgav1. Also add a shell script that will clone and build
libgav1 using the andoroid NDK for the architectures supported on
Android by default.

joedrago commented:
I haven't actually read the change here yet, but just as a note, there is no requirement to build any of the AV1 libraries locally to this library. ideally you would just build libgav1 for the Android system as a shared library using whatever optimizations you want, and then simply let libavif use the system shared library.

If @wantehchang thinks this is a good change, I will defer to him, but the "LOCAL" features are intended to be used in embedded/static builds, which I'm not sure Android qualifies. This feels like a little bit of a misfeature (at first glance).

vigneshvg commented:
The idea here is to let apps bundle libavif if they choose to. For example, an app could write a jni wrapper around libavif and choose to build the necessary native libraries using CMake. This example supports that use case.

There is also a plan to add a JNI wrapper for libavif. Whether or not the JNI wrapper belongs in this repository is another question. But this PR basically lets app build libavif (along with the underlying libraries) from source for multiple architectures if the apps choose to do that. :)

wantehchang commented on ext/libgav1_android.sh:
```diff
@@ -0,0 +1,41 @@
+      -DLIBGAV1_ANDROID_NDK_PATH=${1} \
+      -DLIBGAV1_THREADPOOL_USE_STD_MUTEX=1 \
+      -DANDROID_ABI=${abi}
+    make -j
```
Nit: Is it possible to use ninja instead of make?

vigneshvg replied:
Done.

wantehchang commented on ext/libgav1_android.sh:
```diff
@@ -0,0 +1,41 @@
+
+if [ $# -ne 1 ]; then
+  echo "Usage: ${0} <path_to_android_ndk>"
+else
```
Nit: After the 'echo' statement, should the script just do "exit 1"? Then we won't need the "else".

vigneshvg replied:
Done.

wantehchang commented on android_jni/.gitignore:
```diff
@@ -0,0 +1,15 @@
+*.iml
```
Please remove the second commit (Add Android JNI Wrapper) from this pull request. It should become a separate pull request.

I am not very familiar with GitHub's pull requests, so I don't know what is the best way to do this.

vigneshvg replied:
Adding a commit to my fork seems to update this pull request as well. I was not expecting that. I have reverted that change now. This pull request only contains the one commit now.

vigneshvg commented on CMakeLists.txt:
```diff
@@ -286,7 +286,13 @@ if(AVIF_CODEC_LIBGAV1)
+        else()
+            set(AVIF_ANDROID_PREFIX "")
+        endif()
+        set(LIB_FILENAME "${CMAKE_CURRENT_SOURCE_DIR}/ext/libgav1/build/${AVIF_ANDROID_PREFIX}libgav1${CMAKE_STATIC_LIBRARY_SUFFIX}")
```
Done.

vigneshvg commented:
I believe i have addressed your comments Wan-Teh.

My github skills are quite limited. So hopefully i didn't screw anything up. Please take another look and let me know if you have any more comments.

wantehchang commented on CMakeLists.txt:
```diff
@@ -293,7 +293,12 @@ if(AVIF_CODEC_LIBGAV1)
     if(AVIF_LOCAL_LIBGAV1)
-        set(LIB_FILENAME "${CMAKE_CURRENT_SOURCE_DIR}/ext/libgav1/build/libgav1${CMAKE_STATIC_LIBRARY_SUFFIX}")
+        set (AVIF_LIBGAV1_BUILD_DIR "${CMAKE_CURRENT_SOURCE_DIR}/ext/libgav1/build")
+        # If ${ANDROID_ABI} is set, look for the library under that subdirectory.
```
Is it true that we must pass -DANDROID_ABI=xxx to cmake when we build libavif with the Android NDK? This webpage says ANDROID_ABI is a required argument:

https://developer.android.com/ndk/guides/cmake#android_abi

vigneshvg replied:
Yes that's correct. In the subsequent pull request, where we actually make use of this in libavif's android JNI, that argument will be set by the android gradle build system automatically.

wantehchang reviewed (changes requested):
Vignesh: Thank you for updating the pull request. I have some more comments.

vigneshvg commented on ext/libgav1_android.sh:
```diff
@@ -0,0 +1,42 @@
+# If you want to use a local build of libgav1 to be used with the android NDK,
+# you must clone the libgav1 repo in this directory first, then enable CMake's
+# AVIF_CODEC_LIBGAV1 and AVIF_LOCAL_LIBGAV1 options.
```
Removed.

vigneshvg commented on ext/libgav1_android.sh:
```diff
@@ -0,0 +1,42 @@
+# The git SHA below is known to work, and will occasionally be updated. Feel
+# free to use a more recent commit.
+#
+# cmake and make must be in your PATH.
```
Removed.

vigneshvg commented on ext/libgav1_android.sh:
```diff
@@ -0,0 +1,42 @@
+#
+# Android NDK: https://developer.android.com/ndk/downloads
+#
+# The git SHA below is known to work, and will occasionally be updated. Feel
```
Changed SHA to tag. It refers to the tag of libgav1 and abseil that's being checked out.

jzern commented on ext/libgav1_android.sh:
```diff
@@ -0,0 +1,36 @@
+# This script will build libgav1 for the default ABI targets supported by
```
This is consistent with the other ext/.sh files, but I think it would be better to have a #! on an executable file unless this is meant to be run from another script, then maybe it doesn't need to be directly executable.

vigneshvg replied:
Done.

jzern commented on ext/libgav1_android.sh:
```diff
@@ -0,0 +1,36 @@
+  echo "Usage: ${0} <path_to_android_ndk>"
+  exit
+fi
+git clone -b v0.17.0 --depth 1 https://chromium.googlesource.com/codecs/libgav1
```
you might want to add the same comment here about updating libgav1.cmd

vigneshvg replied:
Done

wantehchang commented on ext/libgav1_android.sh:
```diff
@@ -0,0 +1,39 @@
+
+if [ $# -ne 1 ]; then
+  echo "Usage: ${0} <path_to_android_ndk>"
+  exit
```
Nit: exit 1

because this is an error.

vigneshvg replied:
Done.

wantehchang reviewed (approved):
LGTM. I would like to test this pull request on my Linux computer first.

vigneshvg commented:
> LGTM. I would like to test this pull request on my Linux computer first.

Thanks Wan-Teh.

I have updated the exit code to exit 1.

wantehchang reviewed (approved):
LGTM. I tested this pull request with Android NDK r22b. Android NDK r23b (the latest LTS version) requires a change to libgav1/cmake/toolchains/android.cmake.

After running ext/libgav1_android.sh, I created the build/arm64-v8a directory in libavif and then ran the following cmake command:

`cmake -G Ninja ../.. -DBUILD_SHARED_LIBS=OFF -DAVIF_CODEC_LIBGAV1=ON -DAVIF_LOCAL_LIBGAV1=ON -DCMAKE_TOOLCHAIN_FILE=~/android-ndk-r22b/build/cmake/android.toolchain.cmake -DANDROID_ABI=arm64-v8a`

Then I ran the ninja command to build libavif.a.
